### ISSUE TITLE
Option to filter out imperative benchmarks

### ISSUE THREAD

**Issue by L3P3** [krausest/js-framework-benchmark#1121]
I know it is a bit hard to decide where to draw the line but I would like to filter out frameworks that involve imperative operations like "set the class of element x", as seen in [here](https://github.com/krausest/js-framework-benchmark/pull/991#issuecomment-1019361450). Even if it is not a direct dom manipulation, I (author of _lui_) consider it _cheating_ if you have all that functional/declarative style and then suddenly "help" the framework to make something faster.

I do not deem imperative style an issue per se but I want to be able to rule it out in my projects as much as possible and rely on plain stupid data => view translation. :wink:

**Comment by krausest** [comment]
Not sure I fully understand your request. Those frameworks should have issue #772 added. Is that the case (xaniajs's is marked #772)?
If those are marked with #772 you can deselect those frameworks:
<img width="751" alt="Bildschirmfoto 2022-10-18 um 20 30 22" src="https://user-images.githubusercontent.com/3374055/196514627-bff7e613-9e7c-46b9-9c6f-158f57d969b0.png">

P.S: Seems like I caused a regression in https://github.com/krausest/js-framework-benchmark/commit/d6f4edd612d9b9d8af162d85cc8a7c84c34fdb6e. If you deselect #772, vanillajs isn't counted anymore and scores 0.97. I'll take a look at that.

**Comment by krausest** [comment]
That regression is fixed now.

**Comment by L3P3** [comment]
I mean I want to see only implementations that rely by 100% on data => view mapping, without any imperative operation on the view itself.

**Comment by L3P3** [comment]
I still think we need an issue flag for implementations that imperatively change elements, like reading/writing HTML properties directly instead of having it derived from state.
One key reason to use declarative design is stability and encapsulation. Manipulating the dom directly on call hurts that principle...

**Comment by krausest** [comment]
Thanks - Currently there are no plans to introduce this feature.

**Comment by L3P3** [comment]
I just wanted to add my current view on this issue: It persists to me. In good talks about ui design, its clarified that if you want a complicated web app that is easy to change by multiple people (scalability+stability), you must have a clear separation and atomization.
One actor takes care of a consistent state model. The many ui components should be pure functions without any side effects. Meaning no state must be duplicated.
Every kind of operation that imperatively skips that managed state (for "simplicity" or performance reasons) violates that architecture and opens the doors wide for hard to spot bugs. Big complicated projects with requirement for high stability cannot afford that.
Everyone can use a framework and to cheat in the benchmark, just skip the data-view-principle and directly modify dom or use a framework-provided function to do just that. Its not illegal but I see this as a big reason to just compare apples and apples: The possibility to compare only pure implementations without these one-time patches/hacks.
If I may help with this flags implementation, let me know. I think having more flags to have a better filter precision would be great in general. 😉

**Comment by L3P3** [comment]
Btw, React was invented for that very reason, there is a good talk by the React inventors. It kicked off the web framework hype in the first place.
Big chunk of the implementations here use the approach already common with jQuery 10 years earlier and calling that a framework makes the term washed down.

**Comment by leeoniya** [comment]
@krausest i'll add a vote to re-open this.

i can't find the toggles shown in [your screenshot](https://github.com/krausest/js-framework-benchmark/issues/1121#issuecomment-1282843236) above.

purely-declarative, straightforward implementations are interesting to me, because in most cases, just because you _can_ optimize through manual DOM (or other esoteric knowledge) doesnt mean you _should_ for that extra 3% gain.

the fastest purely-declarative frameworks already land in the 1.05 range. i would love to be able to filter out everything with "Notes" and use the fastest remaining implementation as the baseline.

**Comment by krausest** [comment]
I agree with your preference, but I fear we won't agree on a definition what a declarative framework really is and where the boundary between imperative and declarative lies.
Was react's setState declarative?
Was svelte 4 declarative? Even if you had quirks like arr[0] = 5; arr = arr?
Was angular-nozone with detectChanges so much more imperative?

I also recently noticed that the issue dialog went missing. I'll have to take a look when that happened.

**Comment by leeoniya** [comment]
yeah i think it's fine to just expose an atomic filter on the multiple issue notes.

there is technically another flag i'd like to see: where a framework requires that you use its store/proxy/reactive primitives. frameworks like react and ivi let you use vanilla js objects whereas others require specific hookups.

this is relevant to my use case where data tables are rendered from JSON.parsing huge query responses from the network. for implementations where every row in the parsed response has to be wrapped in a special store primitive. this becomes a required overhead in both memory and cpu. whereas something like react or ivi can simply consume the parsed json at 0 cost.

e.g.
https://github.com/krausest/js-framework-benchmark/blob/3bd194331d0f15cfaa89d97c8864dc7ca9183d2d/frameworks/keyed/solid/src/main.jsx#L12-L21

maybe this point is moot since the existing benchmarks already account for this cost, but i suspect it would not be equivalent if an existing parsed structure had to be wrapped in a store [as a second pass] to optimize the framework diffing cost.

EDIT: since this benchmark is specifically a data table, it may even be more realistic for the data itself to come from the server instead of being generated in memory.

**Comment by krausest** [comment]
I restored the dialog to select which issues are allowed on a new branch. Then I let github copilot repeat the same task. It did quicker and better: https://krausest.github.io/js-framework-benchmark/current.html
Claude Opus decided to also filter vanillajs which was different in my version. The "slowdown" factors still include frameworks that are not selected due to the selected issues (similar to select frameworks).

**Comment by L3P3** [comment]
> I agree with your preference, but I fear we won't agree on a definition what a declarative framework really is and where the boundary between imperative and declarative lies. Was react's setState declarative? Was svelte 4 declarative? Even if you had quirks like arr[0] = 5; arr = arr? Was angular-nozone with detectChanges so much more imperative?
> 
> I also recently noticed that the issue dialog went missing. I'll have to take a look when that happened.

As I said, for me, its about imperative html manipulation.
When the view is derived from state in a purely declarative manner, its fine for me.
But as soon as there is some kind of element.property=value; or innerHTML=`...`, its no more a pure implementation.
And its per implementation, not per framework. You can implement a react app that is pure but you can also add stuff to modify html directly.

**Comment by krausest** [comment]
Ok, then #772 should be pretty fine.

**Comment by L3P3** [comment]
Jein, the big difference is that some implementations use a library function to just do direct dom manipulation.
They dont do `element.innerHTML="<p>blah</p>";` but instead `library.render("<p>$foo</p>", {foo: "blah"}, element);` or similar.
Its not a direct dom manipulation but circumvents the state-view-model the very same way.
Either soften-up #772 to also catch those or we need a new issue instead (this). 😉
Specific example: mikado is a neat library, I like it. But in its example, in https://github.com/nextapps-de/mikado/blob/ab9460460816354dd5a04a6fa265ecc1071f2993/examples/todomvc/src/js/controller.js#L36, if the state in an object is changed, you need to explicitly call it to render the html again from the state object. (A pimped-up way of `innerHTML=someTemplate(yourState)`) So the state and the view is not really managed by the framework, but the library is still used to manually update the html. Sure, its fast, but there is still a single place in the user app code where the state is managed and the library instructed to apply the changes to html. And when you have many components in the app with their own state, you might get into trouble down the road, when multiple places call the `render()` function without knowing about the others.
A true framework should be responsible of owning the single-point-of-truth state and then make sure the view is in sync all the time.
That being said, its hard to draw the line for me too. But there are many cases where #772 does not fit but still its manual dom tweaking on wheels and not appropiate for complex big production apps with a big team like that.

**Comment by L3P3** [comment]
Maybe the new issue name makes more sense. 🙂
#772 is a subset of this.
And true frameworks as introduced by React* are the inverse set of this.

*I heavily recommend this talk: https://youtu.be/XxVg_s8xAms

**Comment by krausest** [comment]
I can add the note it if we agree what frameworks to flag. Would you like to name some non-772-implementations that should get the new flag? Then we'll ask the contributors for their opinion.

**Comment by L3P3** [comment]
I could do a pr and mark all.

**Comment by leeoniya** [comment]
> Jein, the big difference is that some implementations use a library function to just do direct dom manipulation.
> They dont do `element.innerHTML="<p>blah</p>";` but instead `library.render("<p>$foo</p>", {foo: "blah"}, element);` or similar.

i would say what you're showing in mikado is not an example of indirect dom manip as you describe above.

state is never managed by the framework. it is always managed by the user. `setState()` in react components is no different. you are explicitly assigning state and signalling to React that re-rendering needs to happen. the framework still has to resolve how the new state needs to be represented in the dom, including any necessary diffing and dom manip.

if you are not using a reactive store that is hooked into the framework, then this is the way to work with plain object state. you update the state and tell the framework to re-render. i think this comes back to my previous comment, that some frameworks require that your state be absorbed into their reactive store (or immutable object wrappers like immer), otherwise there is no way to manually trigger a re-render. some frameworks expose the low-level primitives that act as `forceUpdate()`, and their reactivity is simply built atop these primitives.

it sounds like we may be asking for the same thing. "does a framework require you to use its integrated store/proxies/signals/immutability for reactivity"?

[1] https://github.com/localvoid/ivi?tab=readme-ov-file#invalidate

<img width="1354" height="1484" alt="Image" src="https://github.com/user-attachments/assets/1c12a2a2-e44e-404e-8e07-a64c35c613ea" />

**Comment by L3P3** [comment]
Good point. And this also touches what I meant with hard to draw the line.

- Having a setState() or dispatch() is separating data from the view. At that point, the developer does not care about dom updates.
- Calling render() to an element is imperatively calling to modify dom.

In my opinion, that line lies between those.

**Comment by leeoniya** [comment]
looks like the store-agnostic vs reactivity convo already happened 6 years ago: https://github.com/krausest/js-framework-benchmark/issues/800#issuecomment-701435064

**Comment by L3P3** [comment]
Yes, there are similarities, but I explicitly draw the line here and want that note. 😉
In short: When I use a framework in production, I dont want to think about when what dom is changed but only care about the state being up to date. If I have to take care of updating the dom, the framework becomes a dom update library.

**Comment by krausest** [comment]
> I could do a pr and mark all.

I'd like to do that as a 2nd step. So far any note caused quite a lot of controversial discussion. I'd like to get into that discussion first 😄.

**Comment by L3P3** [comment]
Fine. But for that, we need an actual controverse. Well, lets try. My provocative thesis: 😉

1. A framework is built around the app, it is the _frame_ around your artwork. You put inside it, how the app should look like. The framework asks you, the app dev, what you want it to do. You tell it, declaratively.
2. However, when you manage the state by yourself and then ask to render a template a with state b into element c, its no more a framework but a dom render library. The app dev has to be careful about conflicts.
3. Thus, true frameworks in the original (React's) sense (see video above) must manage state and dom updating by itself, the app dev just describes to it, on a high level.

Change my mind.

**Comment by krausest** [comment]
Definitions become difficult only when they're applied to examples 😄 
Could you please name a few implementations that are in your opinion #1121, but not #772? Then we'll ask their authors.
